### DISCUSSION TITLE
Form error not updating with Zod refine

### DISCUSSION THREAD

**Discussion by BrendanC23** [shadcn-ui/ui#2120] [Q&A]
I'm running into an issue with React-Hook-Form, Zod, and Shadcn where Zod refine() runs but the error message displayed in `<FormMessage />` does not update.

1. See [this Codesandbox](https://codesandbox.io/p/devbox/nameless-wildflower-95vqmy?file=%2Fapp%2Fpage.tsx)
2. Start typing in the confirm password field.
3. Look at the console - the refine() call is executing and the passwords do not match. However, no error is being shown.
4. Type into the password field. If the passwords do not match, an error will be shown.

Essentially, typing into the password field causes the error message to update, but typing into the confirm password field does not change the error.

https://github.com/shadcn-ui/ui/assets/4711227/bc53949e-a24f-470b-a3b6-3adde9afcad7

**Comment by onurhan1337** [comment]
In the original code, the `refine` method was being used on the entire schema, and it was checking if `data.password` is equal to `data.confirmPassword`. The `data` object here represents all the fields in your form.

In the updated code, the `refine` method is still being used on the entire schema, but the function passed to `refine` has been simplified to a single line: `data => data.password === data.confirmPassword`. This function does exactly the same thing as before: it checks if `data.password` is equal to `data.confirmPassword`.

You can update your `formSchema`:

```js
const FormSchema = z
  .object({
    password: z.string(),
    confirmPassword: z.string(),
  })
  .refine(data => data.password === data.confirmPassword, {
    message: "The passwords must match.",
    path: ["confirmPassword"], // This sets which field the error is attached to
  });
  ```
The second argument to `refine` is an object that specifies the error message to display if the validation fails, and the `path` property in this object specifies which field the error message should be attached to. In the updated code, the error message is attached to the `confirmPassword` field, which means that if the `password` and `confirmPassword` fields don't match, the error message will be displayed next to the `confirmPassword` field.

**Reply by BrendanC23** [reply]
This just has the opposite problem. Now, changing password doesn't update the error. Here's a [Codesandbox](https://codesandbox.io/p/devbox/determined-tdd-gw959d) with your code.

1. Start typing "password" in the confirm field. The error message shows as long as the passwords don't match. Once they match, it disappears.
2. Type additional text into the password field. Now, the passwords no longer match. However, no error is being shown.

Is there a way to get the error messages (regardless of if it's shown in the password field or the confirm field) to be updated when either the password or confirm password fields are changed? The resolver runs when either field is changed, but the error won't update unless that field is in the path of the resolver error.

**Reply by BrendanC23** [reply]
Bump. I'd love to see if anyone has a suggestion to this.

**Comment by ginabeki** [comment]
Going through a similar problem. Would love to see if anyone has a solution to this

**Reply by Enkratia** [reply]
@ginabeki, in case you are still looking for solution: https://stackoverflow.com/a/72166023/19366149

I just added `submitCount` to this solution to avoid checking until form is submitted: 

```
  const {
   // **
    formState: { errors, submitCount },
  } = useForm<InputType>({
   // **
  });
```

```
  const password = watch("password");
  React.useEffect(() => {
    if (!submitCount) return;
    trigger("confirmPassword");
  }, [password, trigger, submitCount]);
```
